<!--
Copyright (c) 2020-2021 Pilz GmbH & Co. KG

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
-->

<!-- Includes starting the scanner driver together with a robot model and preconfigured rviz. -->
<launch>
  <!-- IP-Address of Safety laser scanner -->
  <arg name="sensor_ip" default="192.168.0.10" />

  <!-- Name of the scanner (required to run multiple scanners) -->
  <arg name="tf_prefix" default="laser_1" />

  <!-- Start angle of measurement in radian -->
  <arg name="angle_start" default="$(eval radians(-137.4))" />

  <!-- End angle of measurement in radian -->

  <arg name="angle_end" default="$(eval radians(137.4))" />

  <!-- Publishing of intensities -->
  <arg name="intensities" default="false" />

  <!-- Scan resolution in radian -->
  <arg name="resolution" default="$(eval radians(0.1))" />

  <!-- IP-Address of host machine -->
  <arg name="host_ip" default="auto" />

  <!-- UDP Port on which monitoring frames (scans) should be received -->
  <arg name="host_udp_port_data" default="55115" />

  <!-- UDP Port used to send commands (start/stop) and receive the corresponding replies -->
  <arg name="host_udp_port_control" default="55116" />

  <!-- Set the following to true in order to publish scan data as soon as a UDP packet is ready, instead of waiting for a full scan -->
  <arg name="fragmented_scans" default="false" />

  <!-- Load scanner config file to publish zonesets -->
  <arg name="config_file" default="" />

  <!-- Start rviz -->
  <arg name="rviz" default="true" />

  <!-- Start scanner driver -->
  <include file="$(find psen_scan_v2)/launch/bringup.launch">
    <arg name="sensor_ip" value="$(arg sensor_ip)" />
    <arg name="tf_prefix" value="$(arg tf_prefix)" />
    <arg name="angle_start" value="$(arg angle_start)" />
    <arg name="angle_end" value="$(arg angle_end)" />
    <arg name="intensities" value="$(arg intensities)" />
    <arg name="resolution" value="$(arg resolution)" />
    <arg name="host_ip" value="$(arg host_ip)" />
    <arg name="host_udp_port_data" value="$(arg host_udp_port_data)" />
    <arg name="host_udp_port_control" value="$(arg host_udp_port_control)" />
    <arg name="fragmented_scans" value="$(arg fragmented_scans)" />
  </include>

  <!-- Publish tf frames for the device -->
  <param name="robot_description" command="$(find xacro)/xacro
    '$(find psen_scan_v2)/urdf/example.urdf.xacro' prefix:=$(arg tf_prefix)" />
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />

  <!-- Publish the defined zonesets from the xml configuration -->
  <group if="$(eval config_file != '')">
    <node ns="$(arg tf_prefix)" name="config_server_node" type="config_server_node" pkg="psen_scan_v2">
      <param name="config_file" value="$(arg config_file)" />
      <param name="frame_id" value="$(arg tf_prefix)" />
    </node>
  </group>

  <!-- Start rviz -->
  <group if="$(arg rviz)">
    <node name="rviz" type="rviz" pkg="rviz" args="-d $(find psen_scan_v2)/config/config.rviz" />

    <!-- Visualize active zoneset -->
    <group if="$(eval config_file != '')">
      <node ns="$(arg tf_prefix)" name="active_zoneset_node" type="active_zoneset_node" pkg="psen_scan_v2" />
    </group>
  </group>
</launch>
